### ISSUE TITLE
Possible to run bmc64 in qemu?

### ISSUE THREAD

**Issue by Jammmmm** [randyrossi/bmc64#108]
I've been trying for hours to get bmc64 running in qemu with no success. I can get a Raspberry Pi3 with Raspbian running just fine though.

I've tried every combination of this command:

```
qemu-system-aarch64 -M raspi3 \
-m 1024 \
-serial stdio \
-kernel ./bmc64files/kernel8-32.img \
-append "root=/dev/mmcblk0p1 rootfstype=raw rw terminal=ttyS0 rootwait" \
-dtb ./bmc64files/bcm2710-rpi-3-b.dtb \
-cpu cortex-a53 \
-drive file=./bmc64files/bmc64withfiles.img,if=sd,format=raw
```

Where:
 - kernel8-32.img is the Pi3 bmc64 kernel file
 - bcm2710-rpi-3-b.dtb is the Pi3 dtb file
 - bmc64withfiles.img is the bmc64 image file, with all required files installed (chargen, kernal, etc)

It just sits at a blank screen. I've tried qemu-system-arm, -M raspi2, different versions of the kernel, with and without the "append" parameters (and every combination within it too), with and without the dtb, etc. Nothing.

I've even tried redirecting all output to qemu's stdio with "console=ttyAMA0,115200 console=tty  highres=off console=ttyS0" in the append parameters and adding -nographic. This works for Raspbian but again, nothing for bmc64.

Help!

**Comment by randyrossi** [comment]
Some hardware is not emulated by qemu.  You can try adding enable_serial=true to the cmdline.txt parameters and see how far it's getting.

**Comment by randyrossi** [comment]
BTW: enable_serial=true will use the tx/rx GPIO 14/15 @ 115200 bps for logging.  I'm not sure how you tell qemu to interpret that though.  But there is probably a way.

**Comment by Jammmmm** [comment]
Unfortunately that didn't do anything.

_-append "console=ttyAMA0,115200 console=tty highres=off console=ttyS0" -nographic_ is how I was able to get Raspbian to display everything to stdio so it definitely works. It just doesn't work with qemu.

It's a shame, but I guess any dev I do will mean writing changes to SD and trying on a physical Pi.

**Comment by boras-pl** [comment]
AFAIK raspi3 emulation has no USB support, so default Raspbian has neither working input in a window nor networking. Only serial console. If the emulator has not been finished yet, it is hard to say how the Circle library reacts when it experiences lack of some functionalities. Some hints are there:
https://github.com/rsta2/circle/blob/master/doc/qemu.txt

**Comment by Jammmmm** [comment]
I tried everything there and nothing. I even tried building a minimal circle-stdlib kernel image and it doesn't seem to work either. Everything works fine on a normal Pi though.

Oh well. Thanks for the tips though!

**Comment by Jammmmm** [comment]
I think this issue should be closed.

qemu's Pi emulation is really really flaky and getting Circle and BMC64 to play nice with it is a nightmare for subpar results.

My hope was to be able to debug to screen, or even stdio/stderr/etc. on my dev machine but realistically that's not going to happen anytime soon. Hopefully qemu improves but for now I've found that if anyone's serious about developing for BMC64, you need a USB to Serial cable.

**Comment by pppq** [comment]
In case anyone is interested, I managed to start bmc64 in QEMU 5.2.0 (v5.2.0-11850-g0f27b14b91-dirty) from Stefan Weil's Windows builds: https://qemu.weilnetz.de/w64/2021/

The modifications needed can be found in my fork: https://github.com/randyrossi/bmc64/compare/master...pppq:qemu?expand=1

To summarize: VideoCore emulation for sound, OpenGL ES and dispmanx support is missing, so I have disabled these features and wrote a very rudimentary `FrameBufferLayer` implementation that uses the emulated frame buffer, with "nearest neighbor" interpolation, a hard-coded screen resolution and a timer-based FPS limiter.

Some of the layers are stacked incorrectly, but the VIC-II screen is visible, and the menu also works by and large.

I also had to enable the serial port and set the SD card's first partition for use by default, as command line parsing also seems to have problems (although this should be supported, in theory).

I used the following CLI options:

```
$ qemu-system-arm -M raspi2b -kernel kernel7.img -serial stdio -device usb-kbd -sd sdcard.qcow2
```

It should be possible to compile the C64 build using `make_all.sh pi2`; I have only tried this one, not the other emulated machine types.

![image](https://user-images.githubusercontent.com/8782039/108612237-e107f700-73e6-11eb-9cef-959588a192f4.png)
